pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,335 @@
+// First client in the slice is the primary client.
+// Channel is used to get notifications about what store to use as primary.
+func newMultiClient(cfg multiConfig) *multiClient {
+	c := &multiClient{
+		clients:    cfg.clients,
+		primaryID:  atomic.NewInt32(0),
+		inProgress: map[int]clientInProgress{},
+
+		rateLimiter:      createRateLimiter(cfg.MirrorRateLimit, cfg.MirrorRateBurst),
+		mirroringEnabled: atomic.NewBool(cfg.MirrorEnabled),
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+
+	go c.watchChanges(cfg.MirrorPrefix)
+	if cfg.RuntimeConfigChannel != nil {
+		go c.watchConfigChannel(cfg.RuntimeConfigChannel)
+	}
+
+	return c
+}
+
+func (m *multiClient) watchConfigChannel(configChannel <-chan MultiRuntimeConfig) {
+	for cfg := range configChannel {
+		if cfg.Mirroring != "" {
+			enable := cfg.Mirroring == "true" || cfg.Mirroring == "enable" || cfg.Mirroring == "enabled"
+
+			old := m.mirroringEnabled.Swap(enable)
+			if old != enable {
+				level.Info(util.Logger).Log("msg", "toggled mirroring", "enabled", enable)
+			}
+		}
+
+		if cfg.PrimaryStore != "" {
+			switched, err := m.setNewPrimaryClient(cfg.PrimaryStore)
+			if switched {
+				level.Info(util.Logger).Log("msg", "switched primary KV store", "primary", cfg.PrimaryStore)
+			}
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "failed to switch primary KV store", "primary", cfg.PrimaryStore, "err", err)
+			}
+		}
+	}
+}
+
+func (m *multiClient) getPrimaryClient() (int, kvclient) {
+	v := m.primaryID.Load()
+	return int(v), m.clients[v]
+}
+
+func (m *multiClient) setNewPrimaryClient(store string) (bool, error) {
+	newPrimaryIx := -1
+	for ix, c := range m.clients {
+		if c.name == store {
+			newPrimaryIx = ix
+			break
+		}
+	}
+
+	if newPrimaryIx < 0 {
+		return false, fmt.Errorf("KV store not found")
+	}
+
+	prev := int(m.primaryID.Swap(int32(newPrimaryIx)))
+	if prev == newPrimaryIx {
+		return false, nil
+	}
+
+	// switching to new primary... cancel clients using previous one
+	m.inProgressMu.Lock()
+	defer m.inProgressMu.Unlock()
+
+	for _, inp := range m.inProgress {
+		if inp.client == prev {
+			inp.cancel()
+		}
+	}
+	return true, nil
+}
+
+func (m *multiClient) registerCancelFn(clientID int, fn context.CancelFunc) int {
+	m.inProgressMu.Lock()
+	defer m.inProgressMu.Unlock()
+
+	m.inProgressCnt++
+	id := m.inProgressCnt
+	m.inProgress[id] = clientInProgress{client: clientID, cancel: fn}
+	return id
+}
+
+func (m *multiClient) unregisterCancelFn(id int) {
+	m.inProgressMu.Lock()
+	defer m.inProgressMu.Unlock()
+
+	delete(m.inProgress, id)
+}
+
+// Runs supplied fn with current primary client. If primary client changes, fn is restarted.
+// When fn finishes (with or without error), this method returns given error value.
+func (m *multiClient) runWithPrimaryClient(origCtx context.Context, fn func(newCtx context.Context, primary kvclient) error) error {
+	cancelFn := context.CancelFunc(nil)
+	cancelFnID := 0
+	defer func() {
```

Comment:
Since this logic is also within the loop, you may define a cleanup function and call it from both places, something like:

```
// Call it to ensure context resources are correctly released
func cleanup() {
		if cancelFn != nil {
			cancelFn()
		}
		if cancelFnID > 0 {
			m.unregisterCancelFn(cancelFnID)
		}
	}

	defer cleanup()

	for {
		cleanup()
```

Reply:
You're 100% right. I was just looking at it and thinking the same.